server/src/main/java/org/elasticsearch/cluster/metadata/ResettableValue.java

```diff
@@ -0,0 +1,216 @@
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.cluster.metadata;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.xcontent.ToXContent;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Function;
+
+/**
+ * This class holds a value of type @{param T} that can be in one of 3 states:
+ * - It has a concrete value, or
+ * - It is missing, or
+ * - It is meant to reset any other when it is composed with it.
+ * It is mainly used in template composition to capture the case when the user wished to reset any previous values.
+ * @param <T>
+ */
+public class ResettableValue<T> {
+    private static final ResettableValue<?> RESET = new ResettableValue<>(true, null);
+    private static final ResettableValue<?> UNDEFINED = new ResettableValue<>(false, null);
+    private static final String SKIP_RESET_VALUES = "reset";
+    private static final Map<String, String> SKIP_RESET_VALUES_PARAMS = Map.of(SKIP_RESET_VALUES, "true");
+
+    private final T value;
+    private final boolean isDefined;
+
+    /**
+     * @return the reset state, meaning that this value is explicitly requested to be reset
+     */
+    public static <T> ResettableValue<T> reset() {
+        @SuppressWarnings("unchecked")
+        ResettableValue<T> t = (ResettableValue<T>) RESET;
+        return t;
+    }
+
+    /**
+     * @return the undefined state, meaning that this value has not been specified
+     */
+    public static <T> ResettableValue<T> undefined() {
+        @SuppressWarnings("unchecked")
+        ResettableValue<T> t = (ResettableValue<T>) UNDEFINED;
+        return t;
+    }
+
+    /**
+     * Wraps a value, if the value is null, it returns {@link #undefined()}
+     */
+    public static <T> ResettableValue<T> create(T value) {
+        if (value == null) {
+            return undefined();
+        }
+        return new ResettableValue<>(true, value);
+    }
+
+    private ResettableValue(boolean isDefined, T value) {
+        this.isDefined = isDefined;
+        this.value = value;
+    }
+
+    /**
+     * @return true if the state of this is reset
+     */
+    public boolean shouldReset() {
+        return isDefined && value == null;
+    }
+
+    /**
+     * @return true when the value is defined, either with a concrete value or reset.
+     */
+    public boolean isDefined() {
+        return isDefined;
+    }
+
+    /**
+     * @return the concrete value when that's defined or null
```

Comment:
Probably worth being explicit that this will return null if it's in either the undefined or the should-reset state (that is the behaviour, right?).